File: src/utils/saveImg.py
from io import BytesIO
from mimetypes import guess_extension
from os import path

import requests
from PIL import Image

from .config import config


def download_image(url: str, filename: str, width_thr=200, height_thr=200):
    # 이미지 URL을 다운로드하여 파일로 저장
    response = requests.get(url)
    if response.status_code != 200:
        return None
    try:
        image = Image.open(BytesIO(response.content))
        width, height = image.size
        if width < width_thr or height < height_thr:
            return None

        content_type = response.headers.get("Content-Type", "")
        if (ext := guess_extension(content_type)) is None:  # 확장자를 결정하지 못하는 경우 기본 확장자 설
            ext = ".jpg"  # 기본값으로 '.jpg' 사용

        filename = f"{filename}{ext}"

        with open(path.join(config.media, filename), "wb") as file:
            file.write(response.content)

        return filename
    except Exception as e:
        return None

File: internal/scrapper/util/saveImg.go
package util

import (
	"Scrapper/internal/app"
	"image"
	"io"
	"mime"
	"net/http"
	"os"
	"path"
)

func DownloadImage(url, fileName string, config *app.Config) (string, error) {
	// 1. URL에서 이미지 다운로드
	res, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer res.Body.Close()

	imgConfig, format, err := image.DecodeConfig(res.Body)
	if err != nil {
		return "", err
	}

	// 3. 크기 필터링
	if imgConfig.Width < config.ImageWithThr || imgConfig.Height < config.ImageHeightThr {
		return "", nil
	}

	// 2. Content-Type으로 확장자 결정
	ext := ".jpg" // 기본 확장자
	if exts, _ := mime.ExtensionsByType("image/" + format); len(exts) > 0 {
		ext = exts[0]
	}

	// 3. 파일 생성 및 이미지 저장
	fname := fileName + ext
	fname = path.Join(path.Dir(config.Media), fname)
	file, err := os.Create(fname)
	if err != nil {
		return "", err
	}
	defer file.Close()

	if _, err = io.Copy(file, res.Body); err != nil {
		return "", err
	}

	return fname, nil
}
